### DISCUSSION TITLE
Automatic update with new doctum releases

### DISCUSSION THREAD

**Discussion by ptomulik** [phptailors/docker-doctum#6] [Ideas]
As proposed in  https://github.com/phptailors/docker-doctum/issues/3#issuecomment-752393177, I'd like to consider updating the repository automatically with each new release of [code-lts/doctum](https://github.com/code-lts/doctum).

The current setup of [phptailors/docker-doctum](https://github.com/phptailors/docker-doctum) is that every new git tag  triggers build of [phptailors/doctum](https://hub.docker.com/r/phptailors/doctum) on docker hub. Git tag names are not related to docker image tags:

- docker image tags have form ``X.X-phpY.Y``, where ``X.X`` is major.minor Doctum version, ``Y.Y`` is PHP version,
- git tags are incremented continuously with each new version that has to be built on docker hub.
 
The code for all currently supported ``X.X-phpY.Y`` images is kept in subdirectories named ``X.X/phpY.Y``, the content is generated from top-level ``*.in`` files. With any change to source code, I have to run ``./update.py`` script to regenerate the code. Common configuration file ``config.py`` keeps the common config and the matrix of supported ``X.X``/``Y.Y`` pairs.

The initial questions I have to answer are:

1.  The doctum.phar is downloaded from https://doctum.long-term.support. Will new releases be already there at the moment an action is triggered (I suppose it's some GitHub action of [phptailors/docker-doctum](https://github.com/phptailors/docker-doctum) that will call appropriate hook on my repo)?
2. Is there a way to obtain all released versions of  [code-lts/doctum](https://github.com/code-lts/doctum) to update my matrix accordingly? Ideal solution would include doctum/php compatibility matrix.
3. The list of tags being built on my docker hub repo is "hard-coded there". This will quickly go out of sync with GitHub. How can I cope with this? Build on GitHub + push to docker hub?

@williamdes, ideas?

**Comment by williamdes** [comment]
> Build on GitHub + push to docker hub?

Just link your repository on docker hub and you can configure automatic builds for tags, branches
This works for my GitHub action

I will re-read this soon but my first quick answer is that I need to build an RSS fead to releases and so a released list file will be done at the same time :)

Also you can give me an URL to hook on new releases

**Reply by ptomulik** [reply]
> > Build on GitHub + push to docker hub?
> 
> Just link your repository on docker hub and you can configure automatic builds for tags, branches
> This works for my GitHub action

Well, this is actually what I had to this point. But I wasn't fully satisfied. I had to maintain a list of build contexts manually on DockerHub and this went quickly out of sync with what I had on GitHub. But, in the meantime, I found better solution, which is under development now. :)

> 
> I will re-read this soon but my first quick answer is that I need to build an RSS fead to releases and so a released list file will be done at the same time :)
> 
> Also you can give me an URL to hook on new releases

The hook isn't ready yet, but it's on my TODO list.

**Comment by williamdes** [comment]
Next Doctum release will drop support for PHP 7.1 and add PHAR PHP 8.0 support

**Comment by ptomulik** [comment]
Hello @williamdes. So, here is the plan:

- [x] implement an action which retrieves an array of releases from [upstream repository](https://github.com/docker/build-push-action); couldn't find one so either there is no such action yet (which seems to be strange) or I'm reinventing the wheel,
- [ ] implement a script which accepts above array of [upstream releases](https://github.com/docker/build-push-action/releases) and generates a matrix of ``--build-arg`` values for ``docker build`` runs (build matrix), 
- [ ] if the above script is be generic enough, provide it as a GitHub action,
- [ ] implement a manual workflow which starts on [workflow_dispatch](https://github.blog/changelog/2020-07-06-github-actions-manual-triggers-with-workflow_dispatch/), retrieves list of releases, generates build plan, and uses [docker/build-push-action](https://github.com/docker/build-push-action) to execute the plan. This workflow will automatically expose hook, but it can be also started manually.

The action from the first point is already published, see [ptomulik/github-action-get-releases](https://github.com/ptomulik/github-action-get-releases).

**Reply by williamdes** [reply]
Perfect !
So I have a script to write on Doctum releases website ?
I was intending to generate a RSS feed and a json compatibility file.

I would agree to write a spec to generate such files.
Something generic that any software could use.

**Reply by ptomulik** [reply]
The whole above list is actually my TODO list. After it's completed, the only thing you'll have to do is to call a webhook, that I'll expose. And then magic shall happen. From what I've seen from GitHub docs, I'll have to give some privileges for a GitHub user which shall call the webhook. It's probably @wdes-bot, correct?

**Reply by ptomulik** [reply]
> Perfect !
> So I have a script to write on Doctum releases website ?
> I was intending to generate a RSS feed and a json compatibility file.
> 

The json compatibility file could be added to assets on GitHub releases. That would be great!

> I would agree to write a spec to generate such files.
> Something generic that any software could use.

That would be great, but maybe we should do some research and try to find existing solutions?

**Reply by williamdes** [reply]
> The whole above list is actually my TODO list. After it's completed, the only thing you'll have to do is to call a webhook, that I'll expose. And then magic shall happen. From what I've seen from GitHub docs, I'll have to give some privileges for a GitHub user which shall call the webhook. It's probably @wdes-bot, correct?

I did not know that was linked to accounts :thinking: 
Anyway that would be the right account or maybe @sudo-robot

> The json compatibility file could be added to assets on GitHub releases. That would be great!

Well, why not. But I feel that having it downloadable in a folder of https://doctum.long-term.support/ would be perfect, or is it for your tool to grab it ?

> That would be great, but maybe we should do some research and try to find existing solutions?

Yup, I enabled https://github.com/code-lts/doctum/discussions if you prefer

**Reply by ptomulik** [reply]
> > [...] From what I've seen from GitHub docs, I'll have to give some privileges for a GitHub user which shall call the webhook. It's probably @wdes-bot, correct?
> 
> I did not know that was linked to accounts thinking

IIRC, GitHub API for [workflow_dispatch](https://github.blog/changelog/2020-07-06-github-actions-manual-triggers-with-workflow_dispatch/)  requires the caller to pass token, which has certain privileges. Here are some exempts from GitHub docs:

*To trigger the custom workflow_dispatch webhook event using the REST API, you must send a POST request to a GitHub API endpoint and provide the ref and any required inputs. For more information, see the "Create a workflow dispatch event" REST API endpoint.* [[1](https://docs.github.com/en/free-pro-team@latest/actions/reference/events-that-trigger-workflows#workflow_dispatch)]

*You must authenticate using an access token with the repo scope to use this endpoint. GitHub Apps must have the actions:write permission to use this endpoint. For more information, see "Creating a personal access token for the command line".* [[2](https://docs.github.com/en/free-pro-team@latest/rest/reference/actions#create-a-workflow-dispatch-event)].

This is in fact reasonable, as any workflow can mess with repository, so the GitHub engine has to know who runs it. No matter whether it's started manually, via local events such as ``push`` or ``pull_request`` or via an exposed REST API endpoint.

> Anyway that would be the right account or maybe @sudo-robot

Whoever runs workflows that create releases on [code-lts/doctum](https://github.com/code-lts/doctum) will probably also call my webhook.

> 
> > The json compatibility file could be added to assets on GitHub releases. That would be great!
> 
> Well, why not. But I feel that having it downloadable in a folder of https://doctum.long-term.support/ would be perfect, or is it for your tool to grab it ?

Nothing prevents us from having it in both places. However, the file as release asset on GitHub is what I can find easilly now, as I have the action that retrieves complete release meta-data, including URLs pointing to assets for particular releases. So, make this an asset on GitHub and I have it for free. :)

> 
> > That would be great, but maybe we should do some research and try to find existing solutions?
> 
> Yup, I enabled https://github.com/code-lts/doctum/discussions if you prefer

**Reply by williamdes** [reply]
Okay, perfect I was not thinking about workflows dispatches because they are not "webhooks" but more dispatch endpoints ;p

But in that case I would need to create a workflow on my side that would be triggered on a new tag and then launch a script to call your workflow dispatch.


And yup I agree about the GH artifact, it would be more future proof than a website link.

**Reply by ptomulik** [reply]
> Okay, perfect I was not thinking about workflows dispatches because they are not "webhooks" but more dispatch endpoints ;p
> 

I admit, I'm still learning this whole GH terminology. And I do this only to missuse these terms later.  :)

**Comment by ptomulik** [comment]
Today was one of my rare "Eureka!" days. Story is about that extra asset containing compatibility information (supported PHP versions). And the thought is: *if I have to download any compatibility file, then why the heck do not just go and pull the [https://raw.githubusercontent.com/code-lts/doctum/{tag_name}/composer.json](https://raw.githubusercontent.com/code-lts/doctum/main/composer.json) with ``tag_name`` taken from release metadata?* :)

**Reply by williamdes** [reply]
Also I have the composer.lock that contains a platform lock for phars that I can expose as an asset.

But indeed I use the PHP requirement here to add the lock: https://github.com/code-lts/doctum/blob/ce3de926c98bd267543fc385dab633926f3e5069/scripts/make-release.sh#L133
A more readable version should be better for scripts, I would like everyone to avoid using any of my source files in scripts so I do not break anyone workflow some day ^^

The files are on my TODO list, I could try to do them this week ;p

**Comment by williamdes** [comment]
The RSS feed is out !
https://doctum.long-term.support/releases.rss

**Reply by williamdes** [reply]
And now the compatibility matrix !
https://doctum.long-term.support/versions-platform-compatibility.json

**Comment by williamdes** [comment]
https://github.com/code-lts/doctum/releases/tag/v5.4.0 🚀
